src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelection.js

```diff
@@ -0,0 +1,283 @@
+// @flow
+import React from 'react';
+import {action, autorun, observable} from 'mobx';
+import {observer} from 'mobx-react';
+import {DatagridStore} from 'sulu-admin-bundle/containers';
+import {Overlay, MultiItemSelection} from 'sulu-admin-bundle/components';
+import {translate} from 'sulu-admin-bundle/services';
+import MediaCollection from '../../containers/MediaCollection';
+import CollectionStore from '../../stores/CollectionStore';
+import MediaSelectionStore from './stores/MediaSelectionStore';
+import MediaSelectionItem from './MediaSelectionItem';
+import mediaSelectionStyles from './mediaSelection.scss';
+
+const ADD_ICON = 'plus';
+const MEDIA_RESOURCE_KEY = 'media';
+const COLLECTIONS_RESOURCE_KEY = 'collections';
+
+type Props = {
+    locale: string,
+    value: Array<string | number>,
+    onChange: (ids: Array<string | number>) => void,
+};
+
+@observer
+export default class MediaSelection extends React.PureComponent<Props> {
+    mediaPage: observable = observable();
+    collectionPage: observable = observable();
+    @observable collectionId: ?string | number;
+    @observable mediaDatagridStore: DatagridStore;
+    @observable collectionDatagridStore: DatagridStore;
+    collectionStore: CollectionStore;
+    mediaSelectionStore: MediaSelectionStore;
+    selectedMediaIds: Array<string | number> = [];
+    @observable overlayOpen: boolean = false;
+    overlayDisposer: () => void;
+
+    componentWillMount() {
+        const {
+            value,
+            locale,
+        } = this.props;
+
+        this.mediaSelectionStore = new MediaSelectionStore(value, locale);
+    }
+
+    componentWillUnmount() {
+        this.destroyStores();
+        this.overlayDisposer();
+    }
+
+    @action openMediaOverlay() {
+        this.overlayOpen = true;
+
+        this.overlayDisposer = autorun(this.createStores);
+    }
+
+    @action closeMediaOverlay() {
+        this.overlayOpen = false;
+        this.destroyOverlay();
+    }
+
+    @action setMediaPage(page: number) {
+        this.mediaPage.set(page);
+    }
+
+    @action setCollectionPage(page: number) {
+        this.collectionPage.set(page);
+    }
+
+    @action setCollectionId(id: ?string | number) {
+        this.collectionId = id;
+    }
+
+    destroyOverlay() {
+        this.selectedMediaIds = [];
+        this.collectionId = undefined;
+
+        this.destroyStores();
+        this.overlayDisposer();
+    }
+
+    createStores = () => {
```

Comment:
I think also all of this stuff could go into the separate `MediaSelectionOverlay` component. Would make this file a lot cleaner IMO.